frontend/src/pages/Forum.js

```diff
@@ -1,20 +1,58 @@
-import { useContext, useState } from 'react';
+import { useContext, useEffect } from 'react';
+import { Link } from 'react-router-dom';
+import { Button, List, Typography } from 'antd';
 import { Context } from '../store';
-import { Typography } from 'antd';
+import { updateThreads } from '../store/actions';
 
 const Forum = () => {
   const [state, dispatch] = useContext(Context);
-  const { Title, Text } = Typography;
+  const { Text, Title } = Typography;
+  
+  useEffect(() => {
+    fetch('http://localhost:8080/api/threads').then(res => {
+      return res.json();
+    }).then(async (data) => {
+      await dispatch(updateThreads(data))
+      console.log(state.auth)
+    })
+  }, [])
+
   return(
     <>
-      <Title style={{textAlign: 'center'}}>Forum</Title>
       {
         state.auth.token
           ?
-          <Text>Logged in</Text>
+          <Button>Create Thread</Button>
           :
-          <Text>Logged out</Text>
+          <Text>You must log in to create a thread</Text>
       }
+      <List itemLayout="horizontal" dataSource={state.threads.data} renderItem={thread => (
+        <List.Item>
+          <div style={{ maxWidth: '100%' }}>
+            <div><Link to={"/user/" + thread.author.username}>{thread.author.username}</Link>
+              {thread.createdAt &&
+                <span> Submitted {thread.createdAt}</span>
+              }
+            </div>
+            <Link to={"/thread/" + thread.id}><Title level={5} ellipsis>{thread.title}</Title></Link>
+            <div>{thread.posts.length} Posts
+              {thread.posts.length > 0 && 
+                <span> Last post by&nbsp;
+                  <Link to={"/user/" + thread.posts[thread.posts.length - 1].author.username}>
+                    {thread.posts[thread.posts.length - 1].author.username}
+                  </Link>
+                {thread.posts[thread.posts.length - 1].createdAt &&
+                  <span> at {thread.posts[thread.posts.length - 1].createdAt}</span>
+                }
+                </span>
+              }
+              {/* {state.auth.user.roles.includes('ROLE_MODERATOR') &&
```

Comment:
With edits, I think you could either check if it includes moderator or admin roles (Those will most likely be the only ones with deletion/editing rights for all posts), or if the post's/thread's author id/username matches with the author.

Each thread/post itself could be a seperate component to contain all that data to make it easier for sure!

Reply:
Good job with everything!

Reply:
Oh yeah... there's SO much refactoring to come in follow-ups ;)